src/VisualStudio/Core/Def/Implementation/TableDataSource/AbstractTableDataSource.cs

```diff
@@ -58,40 +69,138 @@ public void Shutdown()
             }
         }
 
-        protected void OnDataRemoved(object key)
+        public abstract ImmutableArray<TableItem<TData>> Deduplicate(IEnumerable<IList<TableItem<TData>>> duplicatedGroups);
+        public abstract ITrackingPoint CreateTrackingPoint(TData data, ITextSnapshot snapshot);
+        public abstract AbstractTableEntriesSnapshot<TData> CreateSnapshot(AbstractTableEntriesSource<TData> source, int version, ImmutableArray<TableItem<TData>> items, ImmutableArray<ITrackingPoint> trackingPoints);
+
+        /// <summary>
+        /// Get unique ID per given data such as DiagnosticUpdatedArgs or TodoUpdatedArgs.
+        /// Data contains multiple items belong to one logical chunk. and the Id represents this particular 
+        /// chunk of the data
+        /// </summary>
+        public abstract object GetItemKey(object data);
+
+        /// <summary>
+        /// Create TableEntriesSource for the given data.
+        /// </summary>
+        public abstract AbstractTableEntriesSource<TData> CreateTableEntriesSource(object data);
+
+        /// <summary>
+        /// Get unique ID for given data that will be used to find data whose items needed to be merged together.
+        /// 
+        /// for example, for linked files, data that belong to same physical file will be gathered and items that belong to
+        /// those data will be de-duplicated.
+        /// </summary>
+        protected abstract object GetOrUpdateAggregationKey(object data);
```

Comment:
So `aggregateKeyMap` basically stores the keys for aggregate data and the value from here will be keyed on the `_map`', I see.